### PR TITLE
Add transaction support

### PR DESCRIPTION
**Pull request by klinson**
requires mongodb version V4.0 or more and deployment replica sets or sharded clusters
transaction support create/insert,update,delete,etc operation
Supports infinite-level nested transactions, but outside transaction rollbacks do not affect the commit of inside transactions
```
DB::beginTransaction();
DB::collection('users')->where('name', 'klinson')->update(['age' => 18]);
DB::transaction(function () {
    DB::collection('users')->where('name', 'mongodb')->update(['age' => 30]);
});
DB::rollBack();
```

### CODE REVIEW

**Comment by Smolevich** [comment]
@klinson check failed tests in travis build please

**Comment by klinson** [comment]
> @klinson check failed tests in travis build please

Because transaction requires mongodb version V4.0 or more and deployment replica sets or sharded clusters, 
But your environment doesn't.

**Comment by klinson** [comment]
@Smolevich Because transaction requires mongodb version V4.0 or more and deployment replica sets or sharded clusters, but your environment doesn't.

**Comment by Smolevich** [comment]
@klinson, Not my environment, it is common configuration based on official docker images, you can change if it's needed

**Comment by Smolevich** [comment]
@klinson if you work on PR, add 'WIP' to title of PR

**Comment by klinson** [comment]
@Smolevich I've modified the environment to support it

File: docker-compose.yml [hunk 1]
```diff
@@ -30,3 +35,46 @@ services:
             - 27017:27017
         logging:
             driver: none
+
+    mongo1:
```
Comment (jenssegers) [inline docker-compose.yml]: Are all of these instances required to build transactions?

Comment (klinson) [reply docker-compose.yml]: Transaction requires mongodb version V4.0 or more and deployment replica sets or sharded clusters, so i must create this environment

Comment (jenssegers) [reply docker-compose.yml]: So mongo has no support for transactions with rollbacks if you're running only 1 instance?

Comment (jenssegers) [reply docker-compose.yml]: That might be confusing for people coming from SQL setups.

**Comment by Smolevich** [comment]
@jenssegers i prepare https://github.com/Smolevich/demo-mongo-application/pull/9 for demo. It is not finished, but today i finish

File: docker-compose.yml (hunk 1 above)
Comment (klinson) [reply docker-compose.yml]: It will report error when you execute `beginTransaction` if the environment is not supported

Comment (Smolevich) [reply docker-compose.yml]: Error [here](https://github.com/Smolevich/demo-mongo-application/pull/9) : `Transaction numbers are only allowed on a replica set member or mongos`

Comment (toocool) [reply docker-compose.yml]: > So mongo has no support for transactions with rollbacks if you're running only 1 instance?

Exactly, I found out about that the hard way, barely found it mentioned in the documentation

Comment (klinson) [reply docker-compose.yml]: see [documentation](https://docs.mongodb.com/manual/core/transactions/#transactions-api)

> - For transactions on MongoDB 4.2 deployments (replica sets and sharded clusters), clients must use MongoDB drivers updated for MongoDB 4.2.
> - When using the drivers, each operation in the transaction must be associated with the session (i.e. pass in the session to each operation).

File: src/Jenssegers/Mongodb/Connection.php [hunk 2]
```diff
@@ -254,4 +260,81 @@ public function __call($method, $parameters)
...
+    }
+
+    /**
+     * get now session if it has session
+     * @return \MongoDB\Driver\Session|null
+     * @author klinson <klinson@163.com>
+     */
+    public function getSession()
+    {
+        return ($this->session_key && isset($this->sessions[$this->session_key]))
```
Comment (Smolevich) [inline src/Jenssegers/Mongodb/Connection.php]:
```suggestion
        return $this->sessions[$this->session_key] ?? null;
```

File: src/Jenssegers/Mongodb/Connection.php [hunk 3]
```diff
@@ -254,4 +260,81 @@ public function __call($method, $parameters)
...
+
+    /**
+     * get now session if it has session
+     * @return \MongoDB\Driver\Session|null
+     * @author klinson <klinson@163.com>
+     */
+    public function getSession()
+    {
+        return $this->sessions[$this->session_key] ?? null;
+            ? $this->sessions[$this->session_key]
```
Comment (Smolevich) [inline src/Jenssegers/Mongodb/Connection.php]:
```suggestion
```

File: src/Jenssegers/Mongodb/Connection.php [hunk 4]
```diff
@@ -254,4 +260,81 @@ public function __call($method, $parameters)
...
+            : null;
```
Comment (Smolevich) [inline src/Jenssegers/Mongodb/Connection.php]:
```suggestion
```

**Comment by Smolevich** [comment]
@klinson do you have plans to add changes for PR?

**Comment by klinson** [comment]
I will continue to change it

**Comment by Smolevich** [comment]
@klinson do you have plans to work on this PR?

**Comment by bsormagec** [comment]
whats your plan about this MR, i need this really. @Smolevich @klinson @jenssegers If you need help or anything i can help to fix this problem i need this immediately. Thanks

**Comment by Giacomo92** [comment]
Merge master, please @klinson

**Comment by klinson** [comment]
How to do? @Giacomo92

**Comment by Giacomo92** [comment]
See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork)

File: MongoReplsetDockerFile [hunk 5]
```diff
@@ -0,0 +1,5 @@
+FROM mongo
```
Comment (Smolevich) [inline MongoReplsetDockerFile]: Add tag please

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,28 @@
+language: minimal
```
Comment (Smolevich) [inline .travis.yml]: Remove travis from PR please

File: src/Jenssegers/Mongodb/Connection.php [hunk 7]
```diff
@@ -20,6 +23,9 @@ class Connection extends BaseConnection
      */
     protected $connection;
 
+    protected $session_key; // sessions会话列表当前会话数组key 随机生成
```
Comment (divine) [inline src/Jenssegers/Mongodb/Connection.php]: Can you please remove comments in Chinese language?

File: src/Jenssegers/Mongodb/Connection.php [hunk 8]
```diff
@@ -20,6 +23,9 @@ class Connection extends BaseConnection
...
+    protected $sessions = []; // 会话列表
```
Comment (divine) [inline src/Jenssegers/Mongodb/Connection.php]: This too.

File: src/Jenssegers/Mongodb/Connection.php [hunk 9]
```diff
@@ -254,4 +260,79 @@ public function __call($method, $parameters)
...
+
+    /**
+     * create a session and start a transaction in session
+     *
+     * In version 4.0, MongoDB supports multi-document transactions on replica sets.
+     * In version 4.2, MongoDB introduces distributed transactions, which adds support for multi-document transactions on sharded clusters and incorporates the existing support for multi-document transactions on replica sets.
+     * To use transactions on MongoDB 4.2 deployments(replica sets and sharded clusters), clients must use MongoDB drivers updated for MongoDB 4.2.
+     *
+     * @see https://docs.mongodb.com/manual/core/transactions/
+     * @author klinson <klinson@163.com>
```
Comment (divine) [inline src/Jenssegers/Mongodb/Connection.php]: I don't think we have author tag for each function.

File: src/Jenssegers/Mongodb/Connection.php [hunk 10]
```diff
@@ -254,4 +260,79 @@ public function __call($method, $parameters)
...
+        $this->sessions[$this->session_key]->startTransaction([
+            'readPreference' => new ReadPreference(ReadPreference::RP_PRIMARY),
+            'writeConcern' => new WriteConcern(1),
+            'readConcern' => new ReadConcern(ReadConcern::LOCAL)
+        ]);
+    }
+
+    /**
+     * commit transaction in this session and close this session
+     * @author klinson <klinson@163.com>
```
Comment (divine) [inline src/Jenssegers/Mongodb/Connection.php]: Remove this, please.

File: src/Jenssegers/Mongodb/Connection.php [hunk 11]
```diff
@@ -254,4 +260,79 @@ public function __call($method, $parameters)
...
+                end($this->sessions);
+                $this->session_key = key($this->sessions);
+            }
+        }
+    }
+
+    /**
+     * get now session if it has session
+     * @return \MongoDB\Driver\Session|null
+     * @author klinson <klinson@163.com>
```
Comment (divine) [inline src/Jenssegers/Mongodb/Connection.php]: Remove this, please.

File: src/Jenssegers/Mongodb/Connection.php [hunk 12]
```diff
@@ -254,4 +260,79 @@ public function __call($method, $parameters)
...
+        if ($session = $this->getSession()) {
+            $session->abortTransaction();
+            $this->setLastSession();
+        }
+    }
+
+    /**
+     * close this session and get last session key to session_key
+     * Why do it ? Because nested transactions
+     * @author klinson <klinson@163.com>
```
Comment (divine) [inline src/Jenssegers/Mongodb/Connection.php]: Remove this, please.

File: src/Jenssegers/Mongodb/Connection.php [hunk 13]
```diff
@@ -254,4 +260,79 @@ public function __call($method, $parameters)
...
+    {
+        if ($session = $this->getSession()) {
+            $session->commitTransaction();
+            $this->setLastSession();
+        }
+    }
+
+    /**
+     * rollback transaction in this session and close this session
+     * @author klinson <klinson@163.com>
```
Comment (divine) [inline src/Jenssegers/Mongodb/Connection.php]: Remove this, please.

**Comment by klinson** [comment]
@divine I will do it today

**Comment by klinson** [comment]
@divine My network can't connect github, I'll try again tomorrow

File: tests/models/User.php [hunk 14]
```diff
@@ -23,7 +23,7 @@ class User extends Eloquent implements AuthenticatableContract, CanResetPassword
 {
     use Authenticatable, CanResetPassword, HybridRelations, Notifiable;
 
-    protected $connection = 'mongodb';
+//    protected $connection = 'mongodb';
```
Comment (Smolevich) [inline tests/models/User.php]: What is the reason to comment this line?

**Comment by Smolevich** [comment]
@klinson, please add another testsuite for transactional tests in phpunit.xml and i prepare new ci file for running only transactional tests

**Comment by Smolevich** [comment]
For example see https://github.com/jenssegers/laravel-mongodb/blob/master/phpunit.xml.dist#L28

**Comment by talnizer** [comment]
@Smolevich Hello, I am a user of jenssegers/laravel-mongodb. As i see currently there is no support of multi doc transaction handling in jenssegers and you people are working on it. Reference to https://github.com/jenssegers/laravel-mongodb/issues/1334 also. Is there any fix 'temporarily' (until jenssegers comes up) to handle this in laravel along with jenssegers and MongoDb/Client because my code is using jenssegers Models at many places in the code. So i don't want to use raw MongoDb/Client alone which required lot of changes and thus, I want to use MongoDB/Client with jenssegers instead of just MongoDB/Client. Any code snippet will be appreciated. Thanks

**Comment by ksgush1** [comment]
Any update on this ? Is this part of the Develop branch already ?

**Comment by divine** [comment]
> Any update on this ? Is this part of the Develop branch already ?

Not part of develop branch. Sorry, still no updates so far.

**Comment by klinson** [comment]
@Smolevich I have added another testsuite for transactional tests in phpunit.xml.What should I do next?

@talnizer @ksgush1 @divine 
Sorry, I stopped coding because of the epidemic.
Now I'm going to keep completing it.

**Comment by Smolevich** [comment]
@klinson i will see on your changes and go back with answer

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 15]
```diff
@@ -600,7 +617,12 @@ public function insert(array $values)
      */
     public function insertGetId(array $values, $sequence = null)
     {
-        $result = $this->collection->insertOne($values);
+        $options = [];
```
Comment (Smolevich) [inline src/Jenssegers/Mongodb/Query/Builder.php:620]: Common code for many methods, better move this logic in special method used in many places

**Comment by Smolevich** [comment]
@klinson can you copy part for ci from this [PR](https://github.com/jenssegers/laravel-mongodb/pull/2019) or i can implemente this functionality at master ?

**Comment by klinson** [comment]
@Smolevich Can I invite you into the project?

File: docker-compose.yml [hunk 16]
```diff
@@ -30,3 +36,62 @@ services:
...
+        logging:
+            driver: none
+
+    mongo_repl_init:
+        image: mongo:4.2.5
+        depends_on:
+            - mongodb_repl
+            - mongodb_repl_2
+            - mongodb_repl_3
+        links:
```
Comment (Smolevich) [inline docker-compose.yml:84]: Delete links please
All containers in same network and can communicate together using service name

File: .github/workflows/build-ci.yml [hunk 17]
```diff
@@ -11,7 +11,7 @@ jobs:
     runs-on: ${{matrix.os}}
     strategy:
       matrix:
-        php: ['7.2', '7.3', '7.4']
+        php: ['7.1', '7.2', '7.3', '7.4']
```
Comment (divine) [inline .github/workflows/build-ci.yml:14]: I think we dropped support for php 7.1

**Comment by divine** [comment]
@klinson well done! Could you please take a loot at @Smolevich https://github.com/jenssegers/laravel-mongodb/pull/1904#discussion_r447489996?

**Comment by klinson** [comment]
Please merge this PR ? @Smolevich

**Review by divine — DISMISSED** [review]

**Comment by tairau** [comment]
Any news?

**Comment by divine** [comment]
> Any news?

There are important problems that needs to be fixed before merging this so no updates planned for now at least.

Thanks!

**Comment by tairau** [comment]
Thanks for the answer. What problems still need to be solved? I see practically solved problems in PR.

**Comment by divine** [comment]
> Thanks for the answer. What problems still need to be solved? I see practically solved problems in PR.

1) Why test are failing?
2) Problems not related to this PR, however a library itself;  see this discussion https://github.com/jenssegers/laravel-mongodb/issues/1974

Thanks!

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 18]
```diff
@@ -344,12 +347,12 @@ public function getFresh($columns = [], $returnLazy = false)
             // Return distinct results directly
             $column = isset($this->columns[0]) ? $this->columns[0] : '_id';
 
+            $options = [];
+            // if transaction in session
+            $options = $this->setSession($options);
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:352]:
```suggestion
            $options = $this->setSession();
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 19]
```diff
@@ -395,6 +398,9 @@ public function getFresh($columns = [], $returnLazy = false)
                 $options = array_merge($options, $this->options);
             }
 
+            // if transaction in session
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:401]:
```suggestion
            // Check if transaction exist in session
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 20]
```diff
@@ -567,7 +573,11 @@ public function insert(array $values)
         }
 
         // Batch insert
-        $result = $this->collection->insertMany($values);
+        $options = [];
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:576]:
```suggestion
```
Please remove.

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 21]
```diff
@@ -567,7 +573,11 @@ public function insert(array $values)
...
+        // if transaction in session
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:577]:
```suggestion
        // Check if transaction exist in session
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 22]
```diff
@@ -567,7 +573,11 @@ public function insert(array $values)
...
+        $options = $this->setSession($options);
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:578]:
```suggestion
        $options = $this->setSession();
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 23]
```diff
@@ -577,7 +587,11 @@ public function insert(array $values)
      */
     public function insertGetId(array $values, $sequence = null)
     {
-        $result = $this->collection->insertOne($values);
+        $options = [];
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:590]:
```suggestion
```
Please remove.

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 24]
```diff
@@ -577,7 +587,11 @@ public function insert(array $values)
...
+        // if transaction in session
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:591]:
```suggestion
        // Check if transaction exist in session
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 25]
```diff
@@ -577,7 +587,11 @@ public function insert(array $values)
...
+        $options = $this->setSession($options);
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:592]:
```suggestion
        $options = $this->setSession();
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 26]
```diff
@@ -598,6 +612,8 @@ public function update(array $values, array $options = [])
         if (!Str::startsWith(key($values), '$')) {
             $values = ['$set' => $values];
         }
+        // if transaction in session
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:615]:
```suggestion
        // Check if transaction exist in session
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 27]
```diff
@@ -620,6 +636,9 @@ public function increment($column, $amount = 1, array $extra = [], array $option
             $query->orWhereNotNull($column);
         });
 
+        // if transaction in session
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:639]:
```suggestion
        // Check if transaction exist in session
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 28]
```diff
@@ -679,7 +698,12 @@ public function delete($id = null)
         }
 
         $wheres = $this->compileWheres();
-        $result = $this->collection->DeleteMany($wheres);
+
+        $options = [];
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:702]:
```suggestion
```

Please remove.

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 29]
```diff
@@ -679,7 +698,12 @@ public function delete($id = null)
...
+        // if transaction in session
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:703]:
```suggestion
        // Check if transaction exist in session
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 30]
```diff
@@ -679,7 +698,12 @@ public function delete($id = null)
...
+        $options = $this->setSession($options);
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:704]:
```suggestion
        $options = $this->setSession();
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 31]
```diff
@@ -704,7 +728,11 @@ public function from($collection, $as = null)
      */
     public function truncate(): bool
     {
-        $result = $this->collection->deleteMany([]);
+        $options = [];
+        // if transaction in session
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:732]:
```suggestion
        // Check if transaction exist in session
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 32]
```diff
@@ -704,7 +728,11 @@ public function from($collection, $as = null)
      */
     public function truncate(): bool
     {
-        $result = $this->collection->deleteMany([]);
+        $options = [];
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:731]:
```suggestion
```

Please remove.

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 33]
```diff
@@ -704,7 +728,11 @@ public function from($collection, $as = null)
...
+        // if transaction in session
+        $options = $this->setSession($options);
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:733]:
```suggestion
        $options = $this->setSession();
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 34]
```diff
@@ -832,6 +860,9 @@ protected function performUpdate($query, array $options = [])
             $options['multiple'] = true;
         }
 
+        // if transaction in session
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:863]:
```suggestion
        // Check if transaction exist in session
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 35]
```diff
@@ -1164,6 +1195,19 @@ public function options(array $options)
         return $this;
     }
 
+    /**
+     * set session for the transaction
+     * @param $session
+     * @return mixed
+     */
+    protected function setSession($options)
```
Comment (divine) [inline src/Jenssegers/Mongodb/Query/Builder.php:1203]:
```suggestion
    protected function setSession($options = [])
```

**Review by divine — CHANGES_REQUESTED** [review]
I'm not sure why you were trying to initialize empty `$options`. It's doing nothing, maybe extra options might be passed - nice idea but not implemented.

**Comment by ymgsmz** [comment]
Any update on transactions support?

File: README.md [hunk 36]
```diff
@@ -960,6 +961,46 @@ If you are familiar with [Eloquent Queries](http://laravel.com/docs/queries), th
...
+DB::collection('users')->where('name', 'klinson')->delete();
+
+// you can commit your changes
+DB::commit();
+
+// you can also rollback them
+//DB::rollBack();
+```
+
+Transaction supports infinite-level nested transactions, but outside transaction rollbacks do not affect the commit of inside transactions.
```
Comment (ymgsmz) [inline README.md:994]:
```suggestion
**NOTE:** Transaction supports infinite-level of nested transactions, but outside transaction rollbacks do not affect the commits of inside transactions.
```

File: src/Jenssegers/Mongodb/Connection.php [hunk 37]
```diff
@@ -21,6 +24,9 @@ class Connection extends BaseConnection
      */
     protected $connection;
 
+    protected $session_key;
+    protected $sessions = [];
```
Comment (ymgsmz) [inline src/Jenssegers/Mongodb/Connection.php:28]: Please add docs

Comment (ymgsmz) [inline src/Jenssegers/Mongodb/Connection.php:28]: Please add docs.

```suggestion
    /**
     * A random unique id for identification of transaction.
     * @var string
     */
    protected $session_key;

    /**
     * A list of transaction sessions.
     * @var array
     */
    protected $sessions = [];
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 38]
```diff
@@ -1164,6 +1195,19 @@ public function options(array $options)
         return $this;
     }
 
+    /**
+     * set session for the transaction
```
Comment (ymgsmz) [inline src/Jenssegers/Mongodb/Query/Builder.php:1199]:
```suggestion
     * Set session for the transaction
```

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 39]
```diff
@@ -1164,6 +1195,19 @@ public function options(array $options)
...
+     * @param $session
+     * @return mixed
+     */
+    protected function setSession($options)
+    {
+        if (!isset($options['session']) && ($session = $this->connection->getSession())) {
+            $options['session']  =  $session;
```
Comment (ymgsmz) [inline src/Jenssegers/Mongodb/Query/Builder.php:1206]:
```suggestion
            $options['session'] = $session;
```

File: src/Jenssegers/Mongodb/Query/Builder.php (hunk 31 above)
Comment (ymgsmz) [reply src/Jenssegers/Mongodb/Query/Builder.php:732]: !! `exists`

**Comment by patie** [comment]
any news?

File: src/Jenssegers/Mongodb/Query/Builder.php [hunk 40]
```diff
@@ -344,12 +347,12 @@ public function getFresh($columns = [], $returnLazy = false)
             // Return distinct results directly
             $column = isset($this->columns[0]) ? $this->columns[0] : '_id';
 
+            $options = [];
```
[the remark this answers is not in the document]
Comment (Smolevich) [reply src/Jenssegers/Mongodb/Query/Builder.php:350]: @klinson can you remove this line?

**Review by Smolevich — COMMENTED** [review]
@klinson see on last comments for your PR

**Comment by Smolevich** [comment]
@klinson copy current ci part . See [here](https://github.com/jenssegers/laravel-mongodb/blob/master/.github/workflows/build-ci.yml#L46)

**Comment by josemiguelq** [comment]
Any news guys?

**Comment by klinson** [comment]
review please @Smolevich

**Comment by mattg66** [comment]
Any updates on this branch being merged, it looks so close to being completed

**Comment by abdulmlik** [comment]
Any updates on this branch

**Comment by redbonzai** [comment]
@jenssegers ... when is this package going to support multi-document transactions?

**Comment by MageloPST** [comment]
Any status on transactions support? Really appreciate this feature to be there. Thanks

**Comment by salvationarinze** [comment]
Any update so far?

**Comment by vinkev** [comment]
Any update with transaction?

File: src/Jenssegers/Mongodb/Connection.php [hunk 41]
```diff
@@ -277,4 +291,78 @@ public function __call($method, $parameters)
...
+     * In version 4.2, MongoDB introduces distributed transactions, which adds support for multi-document transactions on sharded clusters and incorporates the existing support for multi-document transactions on replica sets.
+     * To use transactions on MongoDB 4.2 deployments(replica sets and sharded clusters), clients must use MongoDB drivers updated for MongoDB 4.2.
+     *
+     * @see https://docs.mongodb.com/manual/core/transactions/
+     * @return void
+     */
+    public function beginTransaction()
+    {
+        $this->session_key = uniqid();
+        $this->sessions[$this->session_key] = $this->connection->startSession();
```
Comment (alcaeus) [inline src/Jenssegers/Mongodb/Connection.php:308]: Given that the session from the connection instance is applied automatically to all operations sent through the query builder, I'd refrain from starting multiple sessions. Looking at the [transaction logic used for PDO](https://github.com/illuminate/database/blob/bd88721cabecee43c9ebc4398a9fcedea7f22092/Concerns/ManagesTransactions.php#L134), the first call to `beginTransaction` will start a new transaction, while subsequent only create save points (or no-op if those are not supported). I generally believe this to be a better alternative that should be used instead.

**Review by alcaeus — CHANGES_REQUESTED** [review]
As commented, I would remove hard-coding of transaction options as well as the ability to start multiple transactions in parallel.

In addition to that, this PR is missing the implementation of the `DB::transaction()` helper. While the default logic would work, the `MongoDB\with_transaction` function shipped with the MongoDB driver contains logic designed to avoid re-invoking the callback for errors that only require a second commit command. The retry logic of that function is a little different from the default logic, but that difference should not have any negative effects.

File: docker-compose.yml (hunk 1 above)
Comment (jmikola) [reply docker-compose.yml]: > So mongo has no support for transactions with rollbacks if you're running only 1 instance?

Late to the discussion here, but I'll point out that you can always create a single-member replica set for testing (not recommended for production). Features such as transactions and change streams just depend on there being oplog, which will exist even with a lone primary server.

**Comment by henriquetroiano** [comment]
Any update on this?

**Comment by alcaeus** [comment]
I have an almost-done version of this locally and will aim to have a new pull request up in the coming days. Stay tuned...

**Comment by henriquetroiano** [comment]
> I have an almost-done version of this locally and will aim to have a new pull request up in the coming days. Stay tuned...

Thank you @alcaeus

**Comment by divine** [comment]
Closed in favor of https://github.com/jenssegers/laravel-mongodb/pull/2465, thank you for your work!

**Pull request closed without merging**